### PR TITLE
ARROW-15622: [R] Implement union_all and union for arrow_dplyr_query

### PR DESCRIPTION
**Pull request by wjones127**
This PR adds support for `dplyr::union` and `dplyr::union_all`. Not sure why, but I find I must use the fully qualified name `dplyr::union` or else will get an error.

### CODE REVIEW

**Comment by wjones127** [comment]
Some example usage:

``` r
library(arrow)
#> 
#> Attaching package: 'arrow'
#> The following object is masked from 'package:utils':
#> 
#>     timestamp
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

tab1 <- arrow_table(x = 1:3)
tab2 <- arrow_table(x = 2:4, y = c("a", "b", "c"))

tab1 |>
    mutate(y = "a") |>
    union_all(tab2) |>
    collect()
#> # A tibble: 6 × 2
#>       x y    
#>   <int> <chr>
#> 1     2 a    
#> 2     3 b    
#> 3     4 c    
#> 4     1 a    
#> 5     2 a    
#> 6     3 a

tab1 |>
    mutate(y = "a") |>
    union_all(tab2) |>
    arrange(x, y) |>
    collect()
#> # A tibble: 6 × 2
#>       x y    
#>   <int> <chr>
#> 1     1 a    
#> 2     2 a    
#> 3     2 a    
#> 4     3 a    
#> 5     3 b    
#> 6     4 c

tab1 |>
    mutate(y = "a") |>
    dplyr::union(tab2) |>
    arrange(x, y) |>
    collect()
#> # A tibble: 5 × 2
#>       x y    
#>   <int> <chr>
#> 1     1 a    
#> 2     2 a    
#> 3     3 a    
#> 4     3 b    
#> 5     4 c
```

<sup>Created on 2022-05-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by wjones127** [comment]
The timeouts on RTools mingw seem to be random. I ran locally on mingw64 and was able to get it to build and pass all R tests.

**Comment by nealrichardson** [comment]
> Not sure why, but I find I must use the fully qualified name `dplyr::union` or else will get an error.

Probably namespace collision: 

```
> union
function (x, y) 
{
    u <- as.vector(x)
    v <- as.vector(y)
    unique(c(u, v))
}
<bytecode: 0x1290ae500>
<environment: namespace:base>
```

File: r/tests/testthat/test-dplyr-union.R [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    example_data
+  )
+
+  test_table <- arrow_table(x = 1:10)
+
+  # Union with empty table produces same dataset
+  expect_equal(
+    test_table %>%
+      union_all(test_table$Slice(0, 0)) %>%
+      collect(test_table, as_data_frame = FALSE),
```
Comment (nealrichardson) [inline r/tests/testthat/test-dplyr-union.R:38]:
```suggestion
      compute(),
```

File: r/tests/testthat/test-dplyr-union.R [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    example_data
+  )
+
+  test_table <- arrow_table(x = 1:10)
+
+  # Union with empty table produces same dataset
+  expect_equal(
+    test_table %>%
+      dplyr::union(test_table$Slice(0, 0)) %>%
+      collect(test_table, as_data_frame = FALSE),
```
Comment (nealrichardson) [inline r/tests/testthat/test-dplyr-union.R:64]:
```suggestion
      compute(),
```

**Review by nealrichardson — APPROVED** [review]
A couple of suggestions on the tests but otherwise LGTM, thanks!

**Pull request closed without merging**